
```{r prep, echo=FALSE,warning=FALSE, include=FALSE}

require(tidyr)
require(ggplot2)
require(data.table)
require(openssl)
require(dplyr)
require(optparse)
require(callr)
require(codetools)

codelib<-'/home/jbonnie1/dandd/dev-dandD/lib/'
source(file.path(codelib,'plot_progressive.R'))

outdir<-'/home/jbonnie1/scr4_blangme2/jessica/dandd/scratch/HGSVC2'

```


## Run random orderings for 10 genomes with only SNVs

```{bash, eval=FALSE}
codelib=/home/jbonnie1/dandd/dev-dandD/lib/
outdir=/scratch4/blangme2/jessica/dandd/scratch/HGSVC2
tag=quad
filelist=/data/blangme2/jessica/HGSVC2/consensus_mom_fchild_and_onef_1m1.txt
#_HG0111|
grep -E "(_HG00513|_HG00514|_HG01114|_NA18534_1)" /data/blangme2/jessica/HGSVC2/consensus_fastas.txt > ${filelist}

sketchdir=${outdir}/sketchdb

python ${codelib}/dandd_cmd.py tree -s ${tag} -f ${filelist} -o ${outdir} -k 19 |& tee ${outdir}/${tag}_quad.log

tree_pick=${outdir}/${tag}_7_dashing_dtree.pickle

python ${codelib}/dandd_cmd.py progressive -d ${tree_pick} -n 5040 #-o ${outdir}/${tag}_progu.csv

#../scratch/progressive_union/sketches/HGSVC2/k20/ngen1/allvar_HG00513_1.fasta.gz.w.20.spacing.20.hll
quad_progu5040_7_dashing.csv
${codelib}/clean_abba.sh ${outdir}/${tag}_progu5040_7_dashing.csv > ${outdir}/${tag}_abba.csv 
python ${codelib}/clean_abba.py ${outdir}/${tag}_abba.csv > ${outdir}/${tag}_abba_clean.csv

```



```{r draw_graphs1, cache=TRUE, echo=FALSE, eval=FALSE}
progu <- read.csv('/scratch4/blangme2/jessica/dandd/scratch/HGSVC2/quad_progu5040_7_dashing.csv') %>% mutate(ngenomes=ngen, delta_pos=delta)


tag='quad'
ngenome=7

abba.quad <- clean_abba(progu)

# delta_quad<-read.csv('/scratch4/blangme2/jessica/dandd/scratch/HGSVC2/quad_progu.csv') %>% 
  # rename(delta_pos=delta) #, ngenomes=ngen)
plotProgressiveUnion(species=tag, out = outdir, abba.quad, nshow=10)
## Look at 1-0 or 0-1 for variants on the X Chromosome in vcf, other haplotype might always have 0

```

```{r abba, cache=TRUE}
# abba.quad <- read.csv('/scratch4/blangme2/jessica/dandd/scratch/HGSVC2/quad_abba_clean.csv')
#abba.quad$added <- apply(select(abba.quad,-step), 1, function(data)
 #  names(which(data == 0)))


ggplot(abba.quad) + geom_boxplot(aes(x=as.factor(step),y=deltadelta)) + scale_y_log10()
# abba.quad %>% pivot_longer(contains("HG"),names_to="who",values_to = "value")
```


```{r abba2, cache=TRUE}
ggplot(abba.quad) + geom_boxplot(aes(x=as.factor(step),y=deltadelta, color=as.factor(HG00514_1))) + geom_point(aes(x=as.factor(step),y=deltadelta, color=as.factor(HG00514_1))) + scale_y_log10() 
ggplot(abba.quad) + geom_boxplot(aes(x=as.factor(step),y=deltadelta, color=as.factor(HG00514_2))) + geom_point(aes(x=as.factor(step),y=deltadelta, color=as.factor(HG00514_2))) + scale_y_log10()
```

```{r facet}
abba.quad$added_before <- factor((abba.quad$HG00513_1 == 1) + (abba.quad$HG00513_2 == 1))
dfsumm <- abba.quad %>% group_by(step) %>% summarise(mean=mean(deltadelta), sd=sd(deltadelta), median=median(deltadelta))
dfj <- inner_join(abba.quad, dfsumm)
plot.dfj <- dfj %>% filter((HG00514_1 == 0 | HG00514_2 == 0) & step >= 3 & step <= 5) 
plot.dfj$Haplotype <- ifelse(plot.dfj$HG00514_1 == 0,"Maternal","Paternal")
# ggplot(dfj %>% filter((HG00514_1 == 0 | HG00514_2 == 0) & step >= 2 & step <= 4), aes(y=(deltadelta-mean)/sd, x=added_before, color=added_before)) + geom_jitter(height=0) + facet_grid(step ~ (HG00514_1 == 0)) + theme_bw()

ggplot(plot.dfj, aes(y=(deltadelta-mean)/sd, x=added_before, color=added_before)) + geom_jitter(height=0, size=.25) + facet_grid(step ~ Haplotype) + theme_bw() + ggtitle("Standardized \u394\u03b4 Captures Multi-Way Haplotype Relationships") + labs(color="Number of Mother's\nhaplotypes already added", x="",y="Standardized \u394 \u03b4") 
#+ theme(plot.caption = element_textbox_simple())
# First row shows Stdized Dd for final step of all length 3 permutations where 3rd step is addition of maternal (right column) or paternal (left column) haplotype for daughter. Second and third rows show same relationship for length 4 and 5 permuations. Stdized Dd far below 0 are indicative of closer than average relationships.")

# Title: Stratified by haplotype
# False --> Paternal
# True --> Maternal
# Color --> Number of Mother's haplotypes already added
# Vertical Axis: Standardized Delta Delta
# Supress horizonal title
# Title: Standardized Delta Delta Reveals/Captures Multi-Way Haplotype Relationships 
# Caption: Stdized Dd across permutations conditioned on _1 Maternal .
# First row shows Stdized Dd for final step of all length 3 permutations where 3rd step is addition of maternal (right column) or paternal (left column) haplotype for daughter. Second and third rows show same relationship for length 4 and 5 permuations. Stdized Dd far below 0 are indicative of closer than average relationships.
# 2,3,4 --> 3,4,5
# (NOTE: Confirm _1 always Mat)


```

```{r dedup}
individualsHap=list(HG00514_1="DM", HG00514_2= "DP", HG00513_1 = "M1", HG00513_2 = "M2", NA18534_1 = "UM", 
                 HG01114_1="UF1", HG01114_2="UF2")

individuals=list(HG00514_1="DM", HG00514_2= "DP", HG00513_1 = "M", HG00513_2 = "M", NA18534_1 = "UM", 
                 HG01114_1="UF", HG01114_2="UF")

# create a list containing the prior haplotypes added using their nicknames
plot.dfj$who <- unlist(lapply(plot.dfj$prior, function(x){
  tmp <- lapply(x, function(y){individuals[y]})
  paste0(unlist(tmp),collapse = "_")
}))

# get rid of the duplicate rows (same component haplotypes in a different permutation at the same step)
quad.dedup <- plot.dfj %>% 
  # rowwise() %>% 
  # mutate(flist4=list(sort(flist2))) %>%
  ungroup() %>% 
  select(who, Haplotype,step,delta,deltadelta,mean,sd,added_before,last, contains("HG"), contains("NA")) %>%
  unique()

filter(quad.dedup, step == 5) %>% 
  ggplot(aes(y=deltadelta, x=added_before)) +
  # geom_boxplot(outlier.shape = NA) + 
  geom_jitter(aes(color=as.factor(unlist(who))),height = 0) +
  facet_grid(step ~ Haplotype) + theme_bw()






```




```{r explode, cache=TRUE, eval=FALSE, echo=FALSE}




```


```{r mutate, cache=TRUE, echo=FALSE, eval=FALSE}

```

