
```{r prep, echo=FALSE,warning=FALSE, include=FALSE}

require(tidyr)
require(ggplot2)
require(data.table)
require(openssl)
require(dplyr)
require(optparse)
require(callr)
codelib<-'/home/jessica/dandd/dev-dandD/lib/'
source(file.path(codelib,'plot_progressive.R'))

outdir<-'/scratch4/blangme2/jessica/dandd/scratch/HVSVC2'

```


## Run random orderings for 10 genomes with only SNVs

```{bash}
codedir=/home/jbonnie1/dandd/dev-dandD/lib


tag=HVSVC2
outdir=/scratch4/blangme2/jessica/dandd/scratch/${tag}
cd ${outdir}
filelist=/data/blangme2/jessica/HVSVC2/consensus_fastas_nonrel_female.txt
python ${codedir}/dandd_cmd.py tree -s ${tag} -f ${filelist} -o ${outdir} -k 19
tree_pick=/scratch4/blangme2/jessica/dandd/scratch/HVSVC2/HVSVC2_34_dashing_dtree.pickle
python ${codedir}/dandd_cmd.py progressive -d ${tree_pick} -n 120
ordering_file=/scratch4/blangme2/jessica/dandd/scratch/HVSVC2/sketchdb/HVSVC2_34_orderings.pickle


tag=HVSVC2_snv
outdir=/scratch4/blangme2/jessica/dandd/scratch/${tag}
cd ${outdir}
filelist=/data/blangme2/jessica/HVSVC2/consensus_snv_fastas_nonrel_female.txt
  
python ${codedir}/dandd_cmd.py tree -s ${tag} -f ${filelist} -o ${outdir} -k 19
tree_pick=/scratch4/blangme2/jessica/dandd/scratch/${tag}/${tag}_34_dashing_dtree.pickle
python ${codedir}/dandd_cmd.py progressive -d ${tree_pick} -r ${ordering_file}

tag=HVSVC2_snv_indel
outdir=/scratch4/blangme2/jessica/dandd/scratch/${tag}
cd ${outdir}
filelist=/data/blangme2/jessica/HVSVC2/consensus_snv_indel_fastas_nonrel_female.txt
python ${codedir}/dandd_cmd.py tree -s ${tag} -f ${filelist} -o ${outdir} -k 19
tree_pick=/scratch4/blangme2/jessica/dandd/scratch/HVSVC2_snv_indel/HVSVC2_snv_indel_34_dashing_dtree.pickle
python ${codedir}/dandd_cmd.py progressive -d ${tree_pick} -r ${ordering_file}



```



```{r draw_graphs1, cache=TRUE}
tag='HVSVC2'
outdir=file.path("/scratch4/blangme2/jessica/dandd/scratch",tag)

hvsvc2 <- read.csv(paste0(outdir, "/",tag,"_progu120_34_dashing.csv")) %>% 
   mutate(dataset="SNVs + SVs + Indels") %>%
  mutate(delta_pos=delta, ngenomes=ngen)
plotProgressiveUnion(species=tag, out = outdir, hvsvc2, nshow=5) + 
  geom_line(data=(summarize(group_by(hvsvc2, ngen, dataset), mean=mean(delta_pos))),
            mapping=aes(y=mean,x=ngen))
plotProgressiveUnion(species="HGHVC2", ngenome = ngenome, out = outdir, delta_snv, nshow=1)


```





```{r draw_graphs1, cache=TRUE}
tag='HVSVC2_snv'
outdir=file.path("/scratch4/blangme2/jessica/dandd/scratch",tag)

hvsvc2_snv <- read.csv(paste0(outdir, "/",tag,"_progu120_34_dashing.csv")) %>% 
   mutate(dataset="SNVs Only") %>%
  mutate(delta_pos=delta, ngenomes=ngen)

plotProgressiveUnion(species=tag, out = outdir, hvsvc2_snv, nshow=5) + 
  geom_line(data=(summarize(group_by(hvsvc2_snv, ngen, dataset), mean=mean(delta_pos))),
            mapping=aes(y=mean,x=ngen))

plotProgressiveUnion(species="HGHVC2_snv", out = outdir, hvsvc2_snv, nshow=5)
plotProgressiveUnion(species=tag, out = outdir, hvsvc2_snv, nshow=1)


```


```{r draw_graphs1, cache=TRUE}
tag='HVSVC2_snv_indel'
outdir=file.path("/scratch4/blangme2/jessica/dandd/scratch",tag)

hvsvc2_snv_indel <- read.csv(paste0(outdir, "/",tag,"_progu0_34_dashing.csv")) %>% 
  mutate(dataset="SNVs + Indels") %>%
  mutate(delta_pos=delta, ngenomes=ngen)

plotProgressiveUnion(species=tag, out = outdir, hvsvc2_snv_indel, nshow=5) + 
  geom_line(data=(summarize(group_by(hvsvc2_snv_indel, ngen, dataset), mean=mean(delta_pos))),
            mapping=aes(y=mean,x=ngen))
plotProgressiveUnion(species="HGHVC2_snv_indel", out = outdir, hvsvc2_snv_indel, nshow=5)
plotProgressiveUnion(species="HGHVC2_snv_indel", out = outdir, hvsvc2_snv_indel, nshow=1)


```




```{r plot_across, cache=TRUE}
norder=max(hvsvc2$ordering)
gcount=max(hvsvc2$ngenomes)

# hvsvc2_sum<-hvsvc2 %>% group_by(ngenomes) %>% summarize(HVSVC2=mean(delta_pos))
# hvsvc2_snv_sum<-hvsvc2_snv %>% group_by(ngenomes) %>% summarize(HVSVC2_SNV=mean(delta_pos))
# hvsvc2_snv_indel_sum<-hvsvc2_snv_indel %>% group_by(ngenomes) %>% summarize(HVSVC2_SNV_INDEL=mean(delta_pos))
# summary <- inner_join(hvsvc2_sum, hvsvc2_snv_sum) %>% inner_join(hvsvc2_snv_indel_sum) 

summary<- bind_rows(hvsvc2,hvsvc2_snv,hvsvc2_snv_indel) %>% group_by(ngen, dataset) %>%
  summarize(mean=mean(delta_pos))

summary %>% 
  ggplot() + geom_line(mapping=aes(y=mean, x=ngen, color=dataset)) +
   theme_bw() +
  scale_x_continuous(breaks= scales::pretty_breaks(10)) +
  xlab("Number of Genomes in Set") +
  ylab("Value of \u03b4") +
  ggtitle(label=paste0("Values of \u03b4 over ",norder," orderings of ",gcount," HGSVC2 "," Genomes")) 

allg <-
  # delta %>%
  # mutate(Indicator=ordering <= 6) %>%
  # # filter(ordering %in% c(1:20)) %>%
  # mutate(ngenomes=as.integer(ngenomes),kval=as.factor(kval))  %>%
  ggplot() +
  geom_smooth(data = hvsvc2_snv, aes(y=delta_pos, x=ngenomes, color="SNV only"), formula=y~x,se=TRUE) +
  # geom_smooth(data = delta_snv_sv, aes(y=delta_pos, x=ngenomes, color="SNV + SV"), method='loess',formula=y~x,se = TRUE) +
  geom_smooth(data = hvsvc2, aes(y=delta_pos, x=ngenomes, color="SNVs + Indels + SVs"), method='loess',formula=y~x, se=TRUE) +
  # geom_smooth(data = delta_indel, aes(y=delta_pos, x=ngenomes, color="Indels"), method='loess',formula=y~x, se=TRUE) +
  # geom_smooth(data = delta_sv, aes(y=delta_pos, x=ngenomes, color="SV only"), method='loess',formula=y~x, se=TRUE) +
    geom_smooth(data = hvsvc2_snv_indel, aes(y=delta_pos, x=ngenomes, color="SNVs + Indels"), method='mean',formula=y~x, se=TRUE) +

  #geom_point(data=function(x) subset(x,Indicator),aes(y=delta_pos, x=ngenomes, shape=kval), size=2) +
  # scale_linetype_manual(name=paste0("Fit (",norder," Orderings)"),values=c(2)) +
  # geom_line(data=function(x) subset(x,Indicator),aes(y=delta_pos, x=ngenomes, group=ordering,color=as.factor(ordering)), size=.5) +
  theme_bw() +
  scale_x_continuous(breaks= scales::pretty_breaks(10)) +
  # scale_shape_discrete(name="argmax(k)")+
  # scale_color_discrete(name = "Random Genome Ordering") +
  xlab("Number of Genomes in Set") +
  ylab("Value of \u03b4") +
  ggtitle(label=paste0("Values of \u03b4 over ",norder," orderings of ",gcount," HGSVC2 "," Genomes")) 
#+xlim(c(1,3))
  # guides(color = "none")

allg
```