
1. Download Files from V2 Freeze 4  

```{bash getdata, eval=FALSE}
ml bcftools
ml tabix
ml anaconda
mkdir raw
cd raw
wget http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/HGSVC2/release/v2.0/integrated_callset/variants_freeze4_indel_insdel_alt.vcf.gz
wget http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/HGSVC2/release/v2.0/integrated_callset/variants_freeze4_snv_snv_alt.vcf.gz
wget http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/HGSVC2/release/v2.0/integrated_callset/variants_freeze4_sv_insdel_alt.vcf.gz
wget http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/HGSVC2/release/v2.0/integrated_callset/README_HGSVC_release_v2
wget http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/HGSVC2/release/v2.0/integrated_callset/MANIFEST_HGSVC_release_v2
wget http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/HGSVC2/release/v2.0/integrated_callset/HGSVC2021_v2_TableFields.xlsx
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/HGSVC2/technical/reference/20200513_hg38_NoALT/hg38.no_alt.fa.gz

tabix variants_freeze4_sv_insdel_alt.vcf.gz
abix variants_freeze4_snv_snv_alt.vcf.gz
tabix variants_freeze4_indel_insdel_alt.vcf.gz

cd ..

#mkdir raw
#mv * raw/.
# Sample info obtained from here: https://www.internationalgenome.org/data-portal/data-collection/hgsvc2
# file name is: igsr-human genome structural variation consortium, phase 2.tsv   
# WHICH HAS A COMMA?! Linux auto renamed to the name below
# raw/igsr-human_genome_structural_variation_consortium_phase2.tsv

#make a list of sample names contained in the vcf (35 rows)
bcftools query -l raw/variants_freeze4_snv_snv_alt.vcf.gz > raw/SampleIds.txt
listsamples=$(cat raw/SampleIds.txt)
readarray -t samples <<<$listsamples
echo ${samples[@]}

cat raw/igsr-human_genome_structural_variation_consortium_phase2.tsv | awk '$2=="female"{print $1}' > raw/SampleIds_female.txt
cat raw/igsr-human_genome_structural_variation_consortium_phase2.tsv | awk '$2=="male"{print $1}' > raw/SampleIds_male.txt
echo HG00514_HG00733_NA19240 | sed 's/_/\n/g' > raw/SampleIds_relatives.txt

```


 2. Make combinations of the vcfs  
```{bash step2, eval=FALSE}
mkdir intermediate

bcftools concat -Oz -a -o intermediate/tmp1.vcf.gz raw/variants_freeze4_snv_snv_alt.vcf.gz raw/variants_freeze4_sv_insdel_alt.vcf.gz
bcftools sort -o intermediate/variants_freeze4_snv_sv_insdel_alt.vcf.gz intermediate/tmp1.vcf.gz
tabix -f intermediate/variants_freeze4_snv_sv_insdel_alt.vcf.gz

bcftools concat -Oz -a -o intermediate/tmp2.vcf.gz intermediate/variants_freeze4_snv_sv_insdel_alt.vcf.gz raw/variants_freeze4_indel_insdel_alt.vcf.gz
bcftools sort -o intermediate/variants_freeze4_snv_sv_indel_insdel_alt.vcf.gz intermediate/tmp2.vcf.gz
tabix -f intermediate/variants_freeze4_snv_sv_indel_insdel_alt.vcf.gz

bcftools concat -Oz -a -o intermediate/tmp1.vcf.gz raw/variants_freeze4_snv_snv_alt.vcf.gz raw/variants_freeze4_indel_insdel_alt.vcf.gz
bcftools sort -o intermediate/variants_freeze4_snv_indel_alt.vcf.gz intermediate/tmp1.vcf.gz
tabix -f intermediate/variants_freeze4_snv_indel_alt.vcf.gz

```


3. Split vcfs by sample and use reference to produce FASTA for each sample/haplotype  

```{bash step3, eval=FALSE}
mkdir samplevcf
mkdir consensus
for i in ${samples[@]}; do
bcftools view -s $i -Oz -o samplevcf/allvar_${i}.vcf.gz intermediate/variants_freeze4_snv_sv_indel_insdel_alt.vcf.gz |& tee samplevcf/allvar_${i}.log
tabix -f samplevcf/allvar_${i}.vcf.gz

## NOTE - HAVE CHOSEN ARBITRARY SELECTION OF HAPLOTYPE BASED ON ORDER
bcftools consensus -H1 -o consensus/allvar_${i}_1.fasta -f raw/hg38.no_alt.fa.gz samplevcf/allvar_${i}.vcf.gz |& tee consensus/allvar_${i}_1.log
bcftools consensus -H2 -o consensus/allvar_${i}_2.fasta -f raw/hg38.no_alt.fa.gz samplevcf/allvar_${i}.vcf.gz |& tee consensus/allvar_${i}_2.log
gzip -v consensus/allvar_${i}_?.fasta
done
realpath consensus/*.gz > consensus_fastas.txt
grep -vE "(_HG00514_|_HG00733_|_NA19240_)" consensus_fastas.txt > consensus_fastas_nonrel.txt
grep -vEf raw/SampleIds_male.txt consensus_fastas_nonrel.txt | sort > consensus_fastas_nonrel_female.txt
#grep -vE "(_HG00514_|_HG00733_|_NA19240_)" ${sortprefix}_out.txt > ${sortprefix}_nonrel.txt
#grep -vEf /home/jbonnie1/scr16_blangme2/jessica/data/HVSVC2/raw/SampleIds_male.txt ${sortprefix}_nonrel.txt > ${sortprefix}_nonrel_female.txt

```


```{bash step3b, eval=FALSE}
mkdir consensus_snv_sv
for i in ${samples[@]}; do
bcftools view -s $i -Oz -o samplevcf/snv_sv_insdel_${i}.vcf.gz intermediate/variants_freeze4_snv_sv_insdel_alt.vcf.gz |& tee samplevcf/snv_sv_insdel_${i}.log
tabix -f samplevcf/snv_sv_insdel_${i}.vcf.gz

## NOTE - HAVE CHOSEN ARBITRARY SELECTION OF HAPLOTYPE BASED ON ORDER
bcftools consensus -H1 -o consensus_snv_sv/snv_sv_insdel_${i}_1.fasta -f raw/hg38.no_alt.fa.gz samplevcf/snv_sv_insdel_${i}.vcf.gz |& tee consensus_snv_sv/snv_sv_insdel_${i}_1.log
bcftools consensus -H2 -o consensus_snv_sv/snv_sv_insdel_${i}_2.fasta -f raw/hg38.no_alt.fa.gz samplevcf/snv_sv_insdel_${i}.vcf.gz |& tee consensus_snv_sv/snv_sv_insdel_${i}_2.log
gzip -v consensus_snv_sv/snv_sv_insdel_${i}_?.fasta
done

realpath consensus_snv_sv/*.gz > consensus_snv_sv_fastas.txt
grep -vE "(_HG00514_|_HG00733_|_NA19240_)" consensus_snv_sv_fastas.txt > consensus_snv_sv_fastas_nonrel.txt
grep -vEf raw/SampleIds_male.txt consensus_snv_sv_fastas_nonrel.txt | sort > consensus_snv_sv_fastas_nonrel_female.txt

```

## Make consensus SNVs + Indels  
```{bash step3c,eval=FALSE}
mkdir consensus_snv_indel
for i in ${samples[@]}; do
bcftools view -s $i -Oz -o samplevcf/snv_indel_${i}.vcf.gz intermediate/variants_freeze4_snv_indel_alt.vcf.gz |& tee samplevcf/snv_indel_${i}.log
tabix -f samplevcf/snv_indel_${i}.vcf.gz

## NOTE - HAVE CHOSEN ARBITRARY SELECTION OF HAPLOTYPE BASED ON ORDER
bcftools consensus -H1 -o consensus_snv_indel/snv_indel_${i}_1.fasta -f raw/hg38.no_alt.fa.gz samplevcf/snv_indel_${i}.vcf.gz |& tee consensus_snv_indel/snv_indel_${i}_1.log
bcftools consensus -H2 -o consensus_snv_indel/snv_indel_${i}_2.fasta -f raw/hg38.no_alt.fa.gz samplevcf/snv_indel_${i}.vcf.gz |& tee consensus_snv_indel/snv_indel_${i}_2.log
gzip -v consensus_snv_indel/snv_indel_${i}_?.fasta
done


realpath consensus_snv_indel/*.gz > consensus_snv_indel_fastas.txt
grep -vE "(_HG00514_|_HG00733_|_NA19240_)" consensus_snv_indel_fastas.txt > consensus_snv_indel_fastas_nonrel.txt
grep -vEf raw/SampleIds_male.txt consensus_snv_indel_fastas_nonrel.txt | sort > consensus_snv_indel_fastas_nonrel_female.txt


```

## Make consensus FASTA files for each sample's SNVs ONLY  
```{bash step3d}
mkdir samplevcf
mkdir consensus_snv
for i in ${samples[@]}; do
bcftools view -s $i -Oz -o samplevcf/snv_${i}.vcf.gz raw/variants_freeze4_snv_snv_alt.vcf.gz |& tee samplevcf/snv_${i}.log
tabix -f samplevcf/snv_${i}.vcf.gz

## NOTE - HAVE CHOSEN ARBITRARY SELECTION OF HAPLOTYPE BASED ON ORDER
bcftools consensus -H1 -o consensus_snv/snv_${i}_1.fasta -f raw/hg38.no_alt.fa.gz samplevcf/snv_${i}.vcf.gz |& tee consensus_snv/snv_${i}_1.log
gzip -v consensus_snv/snv_${i}_1.fasta

## Create per sample fasta with second haplotype and then compress
bcftools consensus -H2 -o consensus_snv/snv_${i}_2.fasta -f raw/hg38.no_alt.fa.gz samplevcf/snv_${i}.vcf.gz |& tee consensus_snv/snv_${i}_2.log
gzip -v consensus_snv/snv_${i}_2.fasta

done

realpath consensus_snv/*.gz > consensus_snv_fastas.txt
grep -vE "(_HG00514_|_HG00733_|_NA19240_)" consensus_snv_fastas.txt > consensus_snv_fastas_nonrel.txt
grep -vEf raw/SampleIds_male.txt consensus_snv_fastas_nonrel.txt | sort > consensus_snv_fastas_nonrel_female.txt


```

## Make consensus FASTA files for each sample's INDELs ONLY  
```{bash step3e, eval=FALSE}

mkdir consensus_indel
for i in ${samples[@]}; do
bcftools view -s $i -Oz -o samplevcf/indel_${i}.vcf.gz raw/variants_freeze4_indel_insdel_alt.vcf.gz |& tee samplevcf/indel_${i}.log
tabix -f samplevcf/indel_${i}.vcf.gz

## NOTE - HAVE CHOSEN ARBITRARY SELECTION OF HAPLOTYPE BASED ON ORDER
bcftools consensus -H1 -o consensus_indel/indel_${i}_1.fasta -f raw/hg38.no_alt.fa.gz samplevcf/indel_${i}.vcf.gz |& tee consensus_indel/indel_${i}_1.log
gzip -v consensus_indel/indel_${i}_1.fasta
## Create per sample fasta with second haplotype and then compress
bcftools consensus -H2 -o consensus_indel/indel_${i}_2.fasta -f raw/hg38.no_alt.fa.gz samplevcf/indel_${i}.vcf.gz |& tee consensus_indel/indel_${i}_2.log
gzip -v consensus_indel/indel_${i}_2.fasta
done

realpath consensus_indel/*.gz > consensus_indel_fastas.txt
grep -vE "(_HG00514_|_HG00733_|_NA19240_)" consensus_indel_fastas.txt > consensus_indel_fastas_nonrel.txt
grep -vEf raw/SampleIds_male.txt consensus_indel_fastas_nonrel.txt | sort > consensus_indel_fastas_nonrel_female.txt

```

## Make consensus FASTA files for each sample's STRUCTURAL VARIANTs ONLY  
```{bash step3f, eval=FALSE}

mkdir consensus_sv
for i in ${samples[@]}; do
bcftools view -s $i -Oz -o samplevcf/sv_${i}.vcf.gz raw/variants_freeze4_snv_snv_alt.vcf.gz |& tee samplevcf/sv_${i}.log
tabix -f samplevcf/sv_${i}.vcf.gz

## NOTE - HAVE CHOSEN ARBITRARY SELECTION OF HAPLOTYPE BASED ON ORDER
bcftools consensus -H1 -o consensus_sv/sv_${i}_1.fasta -f raw/hg38.no_alt.fa.gz samplevcf/sv_${i}.vcf.gz |& tee consensus_sv/sv_${i}_1.log
gzip -v consensus_sv/sv_${i}_1.fasta
# create per sample fasta for 2nd haplotype
bcftools consensus -H2 -o consensus_sv/sv_${i}_2.fasta -f raw/hg38.no_alt.fa.gz samplevcf/sv_${i}.vcf.gz |& tee consensus_sv/sv_${i}_2.log
gzip -v consensus_sv/sv_${i}_2.fasta

done

realpath consensus_sv/*.gz > consensus_sv_fastas.txt
grep -vE "(_HG00514_|_HG00733_|_NA19240_)" consensus_sv_fastas.txt > consensus_sv_fastas_nonrel.txt
grep -vEf raw/SampleIds_male.txt consensus_sv_fastas_nonrel.txt | sort > consensus_sv_fastas_nonrel_female.txt


```

```{bash step4}
#mkdir consensus_autosomes
#for i in ${samples[@]}; do
#bcftools view -t^chrX,^chrY -Oz -o samplevcf/allvar_autosomes_${i}.vcf.gz samplevcf/allvar_${i}.vcf.gz |& tee samplevcf/allvar_autosomes_${i}.log
#tabix -f samplevcf/allvar_autosomes_${i}.vcf.gz

## NOTE - HAVE CHOSEN ARBITRARY SELECTION OF HAPLOTYPE BASED ON ORDER
#bcftools consensus -H1 -o consensus_autosomes/allvar_autosomes_${i}_1.fasta -f raw/hg38.no_alt.fa.gz samplevcf/allvar_autosomes_${i}.vcf.gz |& tee consensus_autosomes/allvar_autosomes_${i}_1.log
#bcftools consensus -H2 -o consensus_autosomes/allvar_autosomes_${i}_2.fasta -f raw/hg38.no_alt.fa.gz samplevcf/allvar_autosomes_${i}.vcf.gz |& tee consensus_autosomes/allvar_autosomes_${i}_2.log
#gzip -v consensus_autosomes/allvar_autosomes_${i}_?.fasta
#done
#realpath consensus/*.gz > consensus_fastas.txt
#grep -vE "(_HG00514_|_HG00733_|_NA19240_)" consensus_autosomes_fastas.txt > consensus_autosomes_fastas_nonrel.txt
#grep -vEf raw/SampleIds_male.txt consensus_autosomes_fastas_nonrel.txt | sort > consensus_autosomes_fastas_nonrel_female.txt



```

